### DISCUSSION TITLE
Turn off LED when signal is lost ?

### DISCUSSION THREAD

**Discussion by userXinos** [awawa-dev/HyperSerialEsp8266#39] [Q&A]
Hi, I think I'm missing something, but is it possible to turn off the LEDs when the computer is turned off (ctrl+alt+del)?

- firmware: firmware_WS281x_RGB.bin
- board: d1 mini
- OS: Fedora 39 (Plasma/Wayland)
- Software: HHDR v20.0.0.0beta2, and platform in web: `Linux (V4L2)`

**Comment by awawa-dev** [comment]
Hi
I'm not sure if I understand correctly where is the problem. LEDs should be turned off on HyperHDR exit during the shutdown.

**Reply by userXinos** [reply]
Yes, it sometimes works, but for some reason the system sometimes does not exit hyperHDR, for example if turn it off via the ctrl+alt+del menu

**Reply by awawa-dev** [reply]
Could you test installer from  https://github.com/awawa-dev/HyperHDR/actions/runs/8215461902 it may have been fixed few days ago.

**Reply by userXinos** [reply]
dnf says `hyperhdr-20.0.0.0~Fedora_39~beta2-1.x86_64 is already installed`

**Reply by awawa-dev** [reply]
`dnf remove hyperhdr` first

**Reply by userXinos** [reply]
mmhm, okay.

demo

https://github.com/awawa-dev/HyperSerialEsp8266/assets/40886699/475a119d-e553-482f-80db-09d7eb0c823b

**Comment by awawa-dev** [comment]
You run the app from the console, not service? Try to redirect HyperHDR logs to some file (also you can use --verbose option). Maybe HyperHDR process is simply killed so there is no time to do anything.

**Reply by userXinos** [reply]
> You run the app from the console,

no, I just launched it right after installation for a quick demo. 
It run normally from the settings in the startup applications

[hdr.log](https://github.com/awawa-dev/HyperSerialEsp8266/files/14571900/hdr.log)

**Reply by awawa-dev** [reply]
Seems there is a crash in relatad to Pipewire capturing pw_stream_queue_buffer and that's why the app doesnt close properly the LED driver. Please open an issue in HyperHDR. I'll come back to it maybe during the weekend, I'm not able to code or debug at the moment.

**Reply by awawa-dev** [reply]
maybe it's back https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/3040 ?

**Reply by userXinos** [reply]
ok, is there anything else I can do to solve the problem ? 

Most likely this is an XY problem, but probably a forced shutdown on the controller side if the signal is lost would still be more reliable

**Reply by awawa-dev** [reply]
According to your logs, this happens before HyperHDR receives any signal to exit so there is no way to prevent it. Also from the logs Pipewire doesnt report that the stream is paused which probably happens in the background (unless crash happens before it's output/flush to the console) when you turn off the computer. For now I'm assumg it's HyperHDR fault unless I found out that Pipewire doesnt report the stream status change. Are you able to compile HyperHDR on your unit to test it? There is manual what is necessary for Fedora https://github.com/awawa-dev/HyperHDR/wiki/Compiling-HyperHDR

**Reply by userXinos** [reply]
[hdr-build.log](https://github.com/awawa-dev/HyperSerialEsp8266/files/14573908/hdr-build.log)

*I don’t think this makes any sense, now the error only has one line: `Pipewire: core error reported`*

**Reply by awawa-dev** [reply]
Probably some undefined behavior. As I can see it on my phone `if (!_framePaused)` will need to be added before pw_stream_queue_buffer  in HyperHDR

**Reply by userXinos** [reply]
yes, now it works!

https://github.com/awawa-dev/HyperHDR/blob/fc8357927db85be3f779184339a3ef6a95427334/sources/grabber/pipewire/PipewireHandler.cpp#L716

to 
```cpp
if (newFrame != nullptr && !_framePaused)
```

**Reply by awawa-dev** [reply]
Great! Could you submit a Pull Request (it can be also done online when you fork HyperHDR and create new branch in your repo and edit the file)? Maybe it's good to check the pause also at the beginning of this method and return immediately if it's true. For example `.....if (_pwStream == nullptr || _framePaused).....`

**Reply by userXinos** [reply]
I don’t know for what, but for some reason after installing it as a package via cpack it doesn’t work...

**Reply by awawa-dev** [reply]
always? or randomly? verify build time in 'about' page.

**Reply by awawa-dev** [reply]
there is also second invoke of pw_stream_queue_buffer at the end of the method

**Reply by userXinos** [reply]
> always? or randomly?

I don't understand... It turns off the led normally only the first time after the build (run via bin or as a package) and never does this again

**Reply by awawa-dev** [reply]
Yes, strange. Check if it happens when you abruptly stop pipewire while HyperHDR is running. Probably systemctl stop pipewire.service not sure about Fedora and if pipewire is a system service or it's running in the user space.

**Reply by userXinos** [reply]
![i](https://github.com/awawa-dev/HyperSerialEsp8266/assets/40886699/80fc50c5-b95e-415c-9340-a384a00cc2eb)

(and led freeze)

**Reply by awawa-dev** [reply]
there is something more in logs later? about LED driver?

**Reply by userXinos** [reply]
I go to the web to disable & enable "Enable system capture" after `start pipewire.socket`

**Reply by awawa-dev** [reply]
no, before that. after you've stopper the pipewire was there anything else related to led driver?

**Reply by userXinos** [reply]
No. before stop I specifically pressed enter a few times and after that too

**Reply by awawa-dev** [reply]
there should be more at around ~23:57:50.093  I must finish for today because I can barely keep myself together (flu or something). I will let you know later. Probable there is another issue than that you fixed (Pipewire). Something related to LED driver but I must debug it.
